### PR TITLE
feat: i18n

### LINKED ISSUE #172: Translations
- [ ] Figure out how to do translations
- [ ] Add weblate integration
- [ ] Add them

### PR DESCRIPTION
**Pull request by Alystrasz**
Adds frontend translation.

Closes #172.

---

##### TODOs

- [x] Translate interface
- [x] Add component to select language
- [ ] Translate backend messages (mainly errors)
- [ ] Study automatic translation of natively-English content (mods' descriptions for instance)
- [ ] Weblate integration?

### CODE REVIEW

**Comment by Alystrasz** [comment]
@GeckoEidechse should I translate developer view?

**Comment by GeckoEidechse** [comment]
> @GeckoEidechse should I translate developer view?

While one could acquire in terms of accessibility, I'd say it's not worth it, as

1. it's intended for developers (which generally are more fluent in English anyway), not end-users
2. It's used a lot for PoC implementation and sees more changes then the rest of UI, meaning that translations will often be outdated ^^

**Comment by Alystrasz** [comment]
Agreed!

**Comment by Alystrasz** [comment]
The "new release available" dialog is not triggered by UI, is it?
I can't find the place to translate its content.

Capture d’écran de 2023-02-17 08-00-23

**Comment by Alystrasz** [comment]
I've almost finished translating the entire UI, and I'm wondering if I should submit this PR as-is, and address remaining TODOs in future PRs.
@GeckoEidechse what do you think of the following points?

> Study automatic translation of natively-English content (mods' descriptions, release notes)

> Weblate integration?

**Comment by GeckoEidechse** [comment]
Both of those TODOs will take quite some time to complete so I think it's fine to move both of them to future PRs ^^

**Comment by GeckoEidechse** [comment]
I'll hopefully give it a look tonight. If there's still no review by tomorrow evening, please ping me cause I may have overlooked it ^^"

**Comment by GeckoEidechse** [comment]
Actually, in the meantime, maybe also a few lines to docs on how to add new translations/languages would be nice.

**Comment by Alystrasz** [comment]
> Actually, in the meantime, maybe also a few lines to docs on how to add new translations/languages would be nice.

Addressed in e727ad92333fa18a88c882a1018c9e4c6689b2c4 :)

**Comment by GeckoEidechse** [comment]
So for some reason the UI part doesn't build on my Ubuntu 22.04 laptop

Full log

```
 *  Executing task: npm run tauri dev 


> tauri
> tauri dev

     Running BeforeDevCommand (`cd src-vue && npm run dev`)

> src-vue@0.0.0 dev
> vite

failed to load config from /home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/vite.config.ts
error when starting dev server:
Error: The package "esbuild-linux-64" could not be found, and is needed by esbuild.

If you are installing esbuild with npm, make sure that you don't specify the
"--no-optional" or "--omit=optional" flags. The "optionalDependencies" feature
of "package.json" is used by esbuild to install the correct binary executable
for your current platform.
    at generateBinPath (/home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/esbuild/lib/main.js:1819:15)
    at esbuildCommandAndArgs (/home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/esbuild/lib/main.js:1886:33)
    at ensureServiceIsRunning (/home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/esbuild/lib/main.js:2051:25)
    at build (/home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/esbuild/lib/main.js:1942:26)
    at bundleConfigFile (file:///home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/vite/dist/node/chunks/dep-5605cfa4.js:63043:26)
    at loadConfigFromFile (file:///home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/vite/dist/node/chunks/dep-5605cfa4.js:63019:31)
    at resolveConfig (file:///home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/vite/dist/node/chunks/dep-5605cfa4.js:62643:34)
    at createServer (file:///home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/vite/dist/node/chunks/dep-5605cfa4.js:61943:26)
    at CAC.<anonymous> (file:///home/gecko/Documents/Coding/Northstar-stuff/Tools/FlightCore/src-vue/node_modules/vite/dist/node/cli.js:707:30)
       Error The "beforeDevCommand" terminated with a non-zero status code.

 *  The terminal process "/usr/bin/bash '-c', 'npm run tauri dev'" terminated with exit code: 1. 
 ```

**Comment by GeckoEidechse** [comment]
I also have issues getting persistence working rn but that seems to be caused by something outside this PR?

File: docs/DEVELOPMENT.md [hunk 1]
```diff
@@ -178,6 +178,26 @@ struct User {
 
 then simply run `cargo test`. The generated bindings are placed in `src-tauri/bindings/`. Make sure to add and commit them as well!
 
+### Internationalization
+
+For FlightCore to be used by the largest number, its interface is translated in several languages; users can choose used language through FlightCore settings.
+
+Localization files are located in `src-vue/src/i18n/lang`.
```
Comment (GeckoEidechse) [inline docs/DEVELOPMENT.md:185]: Would also be nice to know how to add an additional language. Is it just adding a file with the right name, e.g. `de.ts` for German or is there more to it? ^^

File: src-vue/src/components/ModsMenu.vue [hunk 2]
```diff
@@ -7,20 +7,20 @@
...
             <!-- Search inputs -->
-            <h5>Filter</h5>
-            <el-input v-model="$store.state.search.searchValue" placeholder="Search" clearable />
+            <h5>{{ $t('mods.menu.filter') }}</h5>
+            <el-input v-model="$store.state.search.searchValue" :placeholder="$t('mods.menu.search')" clearable />
             <el-select
                 v-if="!showingLocalMods"
                 v-model="$store.state.search.sortValue" 
-                placeholder="Sort mods"
+                :placeholder="$t('mods.menu.sort_mods')"
```
Comment (GeckoEidechse) [inline src-vue/src/components/ModsMenu.vue:23]: remind me, what does the leading `:` entail? ^^

File: src-vue/src/i18n/lang/fr.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
+export default {
+    menu: {
```
Comment (GeckoEidechse) [inline src-vue/src/i18n/lang/fr.ts:2]: What happens if a language is missing a key. Does it fall back to English?

File: src-vue/src/components/LanguageSelector.vue [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import { Store } from 'tauri-plugin-store-api';
+const persistentStore = new Store('flight-core-settings.json');
+
+export default defineComponent({
+    name: 'LanguageSelector',
+    data: () => ({
+        value: '',
+        options: [
+            {
+                value: 'en',
```
Comment (GeckoEidechse) [inline src-vue/src/components/LanguageSelector.vue:26]: Shouldn't it technically be `fr-FR` and `en-UK` (I think I put most stuff in UK spelling) to allow for variations of a language, e.g. UK vs US or Spanish vs Mexican Spanish?

File: src-vue/src/plugins/store.ts [hunk 5]
```diff
@@ -313,8 +313,9 @@ export const store = createStore<FlightCoreStore>({
 
             // Display notification to highlight change
             ElNotification({
+                // TODO translate
```
Comment (GeckoEidechse) [inline src-vue/src/plugins/store.ts:316]: Leftover TODO? ^^

File: docs/DEVELOPMENT.md [hunk 6]
```diff
@@ -178,6 +178,26 @@ struct User {
...
+    {{ $t('menu.play') }}
+</div>
+```
+
+For use in Typescript code, invoke the `this.$t` method:
+
+```javascript
+return this.$t("play.button.select_game_dir");
+```
+
```
Comment (GeckoEidechse) [inline docs/DEVELOPMENT.md:200]: Maybe also mention how to integrate some variable string into a translation like for example for switching release channel:

```ts
message: i18n.global.tc('channels.release.switch.text', {canal: state.northstar_release_canal}),
```
```ts
    channels: {
        release: {
            switch: {
                text: "Switched release channel to {canal}."
            }
        }
    }
```

File: src-vue/src/views/RepairView.vue [hunk 7]
```diff
@@ -1,30 +1,30 @@
 <template>
     <div class="fc-container">
         <el-scrollbar>
-            <el-alert title="Info" type="info" :closable="false" show-icon>
-                This window contains various functionality to repair common issues with Northstar and FlightCore.
+            <el-alert title="Information" type="info" :closable="false" show-icon>
```
Comment (GeckoEidechse) [inline src-vue/src/views/RepairView.vue:4]: Should that line also be translated?

**Review by GeckoEidechse — COMMENTED** [review]
Didn't manage to get through the whole thing yet so here's an initial review of what I have so far ^^

File: src-vue/src/components/ModsMenu.vue (hunk 2 above)
Comment (Alystrasz) [reply src-vue/src/components/ModsMenu.vue:23]: `placeholder="Sort mods"` uses the `"Sort mods"` as-is, while adding a leading `:` means the content of `""` will be interpreted as javascript code (here, injecting translation into selector component)

File: src-vue/src/i18n/lang/fr.ts (hunk 3 above)
Comment (Alystrasz) [reply src-vue/src/i18n/lang/fr.ts:2]: Yep, it defaults to English, as set in `main.ts`:
```javascript
export const i18n = createI18n({
    locale: 'en',
    fallbackLocale: 'en',
    messages: {
        en, fr
    }
});
```

File: docs/DEVELOPMENT.md (hunk 6 above)
Comment (Alystrasz) [reply docs/DEVELOPMENT.md:213]: Added in 333a7c2f3c3aec0b867562a28a2ba024dc7a0c01

File: src-vue/src/plugins/store.ts (hunk 5 above)
Comment (Alystrasz) [reply src-vue/src/plugins/store.ts]: Nah, I forgot to translate those :)
Fixed in 50161e678a96abdf233f867e5640184c15d9ee7a

File: src-vue/src/views/RepairView.vue (hunk 7 above)
Comment (Alystrasz) [reply src-vue/src/views/RepairView.vue:4]: Which line are you talking about?

File: src-vue/src/components/LanguageSelector.vue (hunk 4 above)
Comment (GeckoEidechse) [reply src-vue/src/components/LanguageSelector.vue:26]: So turns out the one I meant is [RFC 5646](https://www.rfc-editor.org/rfc/rfc5646.html). Anyway, ISO-3166 is probably fine for now ^^

File: src-vue/src/views/RepairView.vue (hunk 7 above)
Comment (Alystrasz) [reply src-vue/src/views/RepairView.vue]: You're right, addressed in dc468a921b84ab259cfa21b12116e8cb71c7fd5b

**Comment by GeckoEidechse** [comment]
So far UI stuff looks good in testing. Will have to give the code another look than it should be good to merge :D

In the meantime, do you maybe wanna update branch with main again ^^

File: src-vue/src/plugins/store.ts [hunk 8]
```diff
@@ -16,8 +16,8 @@ import { ReleaseInfo } from "../../../src-tauri/bindings/ReleaseInfo";
 import { ThunderstoreMod } from "../../../src-tauri/bindings/ThunderstoreMod";
 import { NorthstarMod } from "../../../src-tauri/bindings/NorthstarMod";
 import { searchModule } from './modules/search';
+import { i18n } from '../main';
 import { pullRequestModule } from './modules/pull_requests';
-import { PullsApiResponseElement } from "../../../src-tauri/bindings/PullsApiResponseElement";
```
Comment (GeckoEidechse) [inline src-vue/src/plugins/store.ts:20]: Technically removing unused library should happen in separate PR but whatever ig ¯\\\_(ツ)\_/¯

File: src-vue/src/components/LanguageSelector.vue [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+            {
+                value: 'fr',
+                label: 'Français'
+            }
+        ]
+    }),
+    mounted: async function() {
+        const lang: string = await persistentStore.get('lang') as string;
+        this.value = lang;
+    },
```
[the remark this answers is not in the document]
Comment (GeckoEidechse) [reply src-vue/src/components/LanguageSelector.vue:38]: Got a fix in #227. I'd say we get the fix merged into `main`, then update this branch to `main` and then do final test and merge ^^

File: src-vue/src/components/ThunderstoreModCard.vue [hunk 10]
```diff
@@ -200,11 +200,11 @@ export default defineComponent({
 
             // Show pop-up to confirm delete
             ElMessageBox.confirm(
-                'Delete Thunderstore mod?',
-                'Warning',
+                this.$t('mods.card.remove_dialog_text'),
+                this.$t('mods.card.remove_dialog_title'),
                 {
                     confirmButtonText: 'OK',
```
[the remark this answers is not in the document]
Comment (Alystrasz) [reply src-vue/src/components/ThunderstoreModCard.vue]: Addressed in 37922f49f5b5076be99cf01b42563904568c0628

File: src-vue/src/views/mods/LocalModsView.vue [hunk 11]
```diff
@@ -1,23 +1,25 @@
...
         <div>
-            <p v-if="mods.length === 0">No mods were found.</p>
+            <p v-if="mods.length === 0">{{ $t('mods.local.no_mods') }}</p>
             <el-card v-else shadow="hover" v-for="mod in mods" v-bind:key="mod.name">
                 <el-switch style="--el-switch-on-color: #13ce66; --el-switch-off-color: #8957e5" v-model="mod.enabled"
                             :before-change="() => updateWhichModsEnabled(mod)" :loading="global_load_indicator" />
                 <el-popconfirm
-                    title="Are you sure to delete this mod?"
+                    :title="$t('mods.local.delete_confirm')"
                     @confirm="deleteMod(mod)"
```
[the remark this answers is not in the document]
Comment (Alystrasz) [reply src-vue/src/views/mods/LocalModsView.vue]: Addressed in 37922f49f5b5076be99cf01b42563904568c0628

File: src-vue/src/components/LanguageSelector.vue [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    methods: {
+        onChange(value: string) {
+            this.$root!.$i18n.locale = value;
+            persistentStore.set('lang', value);
```
Comment (GeckoEidechse) [inline src-vue/src/components/LanguageSelector.vue:42]: make sure to also save afterwards with `await persistentStore.save();` as in #227

**Review by GeckoEidechse — APPROVED** [review]
Code looks fine as I can tell and it works fine in testing.
All requested changes have been addressed.
Language saves in persistence as expected now.

**Pull request merged**